keps/sig-storage/20190204-scale-testing.md

```diff
@@ -0,0 +1,215 @@
+authors:
+  - "@msau42"
+owning-sig: sig-storage
+participating-sigs:
+  - sig-scalability
+reviewers:
+  - "@pohly"
+  - "@saad-ali"
+  - "@wojtek-t"
+approvers:
+  - "@saad-ali"
+  - "@wojtek-t"
+editor: TBD
+creation-date: 2019-02-04
+last-updated: 2019-02-04
+status: provisional
+see-also:
+- https://github.com/kubernetes/community/blob/master/sig-scalability/slos/pod_startup_latency.md
+- TODO link to stateful pod SLI
+- https://github.com/kubernetes/perf-tests/blob/master/clusterloader2/docs/design.md
+replaces:
+superseded-by:
+---
+
+# Volume Scale and Performance Testing Plan
+
+## Table of Contents
+
+* [Summary](#summary)
+* [Motivation](#motivation)
+   * [Goals](#goals)
+   * [Non-Goals](#non-goals)
+* [Proposal](#proposal)
+   * [WIP Test Environment](#wip-test-environment)
+   * [Test Cases](#test-cases)
+      * [Pod Startup](#pod-startup)
+      * [Pod Teardown](#pod-teardown)
+      * [WIP: PV Binding Tests](#wip-pv-binding-tests)
+      * [WIP: PV Provisioning Tests](#wip-pv-provisioning-tests)
+      * [WIP: PV Deletion Tests](#wip-pv-deletion-tests)
+* [Graduation Criteria](#graduation-criteria)
+   * [Phase 1](#phase-1)
+   * [Phase 2](#phase-2)
+* [Implementation History](#implementation-history)
+
+
+## Summary
+
+This KEP outlines a plan for testing scalability and performance of K8s storage components.
+
+## Motivation
+
+Adding storage scale and performance tests will help:
+* Understand the current scale limits of the Kuberentes storage system.
+* Set expectations (SLOs) for consumers of the Volume API.
+* Determine bottlenecks and influence which need addressing.
+
+### Goals
+
+* Measure the overhead of K8s components owned by sig-storage:
+  * K8s volume controllers
+  * Kubelet volume manager
+  * CSI sidecars
+* Stress various dimensions of volume operations to determine:
+  * Max volumes per pod
+  * Max volumes per node
+  * Max volumes per cluster
+* Test with the following volume types:
+  * EmptyDir
+  * Secret
+  * Configmap
+  * Downward API
+  * CSI mock driver
+  * TODO: Hostpath?
+  * TODO: Local?
+* Provide tests that vendors can easily run against their volume drivers.
+
+### Non-Goals
+
+* Test and measure storage provider’s drivers.
+
+## Proposal
+
+### Test Framework
+
+The tests should be developed and run in sig-scalability’s test
+[Cluster Loader framework](https://github.com/kubernetes/perf-tests/blob/master/clusterloader2/README.md)
+and infrastructure.
+
+The framework already supports measurements for Pod startup latency and
+scheduling metrics. More measurements can be added to get finer-grained details
+for volume operations:
+
+* Gather [volume operation
+  metrics](https://github.com/kubernetes/kubernetes/blob/master/pkg/volume/util/metrics.go)
+  from kube-controller-manager
+  * Caveat: These metrics measure the time of a single operation attempt and not
+    the e2e time across retries. These metrics may not be suitable to measure
+    latency from a pod's perspective.
```

Comment:
I think that we need to clearly distinguish two things:
(a) the metrics that are implementing SLI 
(b) metrics that are useful for understanding better what is happening and/or debugging failures

Most of the questions here are for (b) category, which is not something that is the most important thing for scalability SIG - this is important for individual SIGs to debug their feature. And how exactly this will be done in their components is out of our responsibility (it is much more in scope of sig instrumentation).
The main two categories of options are:
- expose a set of aggregated metrics from your component (and then we can provide mechanisms for gathering them at the framework level - the problem isn't fully solved in ClusterLoader, but it's something we are looking at as this is really needed)
- tract this information in a more custom way in your component by:
  - some kind of tracing (until we have it, maybe traces in logs: https://github.com/kubernetes/utils/tree/master/trace are good enough for this purpose)
  - extending objects to keep more information (though we should increase amount of operations on apiserver due to that reason)
But this is more individual - there isn't one single good option that will work in all cases.

@pohly @msau42

Reply:
Agree.  For pod startup SLI, we already have the necessary metrics in place.  

We already have metrics for volume operations, but like I mentioned above, they're not really good for measuring e2e time to successfully complete the operation.  So the options I see are:

* Add new metrics that can track total e2e time for the operation. This was attempted already in https://github.com/kubernetes/kubernetes/pull/57766, but it got complicated when we tried to handle operation cancellations/complete failure.
* Add tracing. Needs investigation if this could be leveraged for more than just scale tests without too much overhead to the system. I still think a "total latency" metric would be very useful beyond scale tests.
* Add timestamps to API objects. This would need to be carefully considered so that we don't trigger additional API updates. Ideally we already have some sort of status that we are already updating when the operation is complete. For something like attach, that can also be detached and reattached, we would need to make sure the timestamp makes sense in each of those states.

Reply:
Also, I'll update the proposal to distinguish these two types of metrics

Reply:
> * I still think a "total latency" metric would be very useful beyond scale tests.

+1

> * Ideally we already have some sort of status that we are already updating when the operation is complete. For something like attach, that can also be detached and reattached, we would need to make sure the timestamp makes sense in each of those states.

Yes - that makes perfect sense to me.